File: windows/Core/Network/UdpKmStream.cs
using BarelyReal.Core.Protocol;
using System.Net;
using System.Net.Sockets;

namespace BarelyReal.Core.Network;

/// UDP socket carrying encoded KmFrame events.
/// TODO: wrap frames in AEAD once TlsSession exposes exporter-derived key material.
public sealed class UdpKmStream : IDisposable
{
    public event Action<KmFrame>? OnFrame;
    public event Action<string>? LogLine;

    private UdpClient? _client;
    private CancellationTokenSource? _cts;
    private UdpKmSourceFilter _sourceFilter = UdpKmSourceFilter.Disabled;
    private ulong _droppedFrames;
    private DateTime _lastDropLogUtc = DateTime.MinValue;

    public void Bind(ushort localPort, string? expectedPeerHost = null)
    {
        Close();

        _sourceFilter = UdpKmSourceFilter.FromHost(expectedPeerHost);
        _droppedFrames = 0;
        _lastDropLogUtc = DateTime.MinValue;
        _client = new UdpClient(new IPEndPoint(IPAddress.Any, localPort));
        _cts = new CancellationTokenSource();
        if (_sourceFilter.Enabled)
            LogLine?.Invoke($"UDP source filter enabled for {_sourceFilter.Description}");
        _ = ReceiveLoopAsync(_client, _cts.Token);
    }

    public void Send(KmFrame frame, string peerHost, ushort peerPort)
    {
        var client = _client ??= new UdpClient();
        var payload = KmFrameCodec.Encode(frame);
        client.Send(payload, payload.Length, peerHost, peerPort);
    }

    public void Close()
    {
        _cts?.Cancel();
        _client?.Dispose();
        _cts?.Dispose();

        _client = null;
        _cts = null;
    }

    public void Dispose()
    {
        Close();
    }

    private async Task ReceiveLoopAsync(UdpClient client, CancellationToken token)
    {
        try
        {
            while (!token.IsCancellationRequested)
            {
                var result = await client.ReceiveAsync(token).ConfigureAwait(false);
                if (!_sourceFilter.Allows(result.RemoteEndPoint.Address))
                {
                    LogDroppedFrame(result.RemoteEndPoint);
                    continue;
                }

                LogLine?.Invoke(
                    $"UDP frame received raw bytes len={result.Buffer.Length} from={result.RemoteEndPoint} hex={PreviewHex(result.Buffer)}");
                try
                {
                    OnFrame?.Invoke(KmFrameCodec.Decode(result.Buffer));
                }
                catch (BrpCodecException ex)
                {
                    LogLine?.Invoke($"UDP frame decode failed: {ex.Error}");
                    // Malformed UDP datagrams are dropped. Control channel owns reconnect policy.
                }
            }
        }
        catch (OperationCanceledException)
        {
            // Expected during Close().
        }
        catch (ObjectDisposedException)
        {
            // Expected during Close().
        }
    }

    private void LogDroppedFrame(IPEndPoint remoteEndPoint)
    {
        _droppedFrames++;
        var now = DateTime.UtcNow;
        if (_droppedFrames <= 3 || _droppedFrames % 100 == 0 || now - _lastDropLogUtc >= TimeSpan.FromSeconds(30))
        {
            _lastDropLogUtc = now;
            LogLine?.Invoke(
                $"UDP frame dropped from unexpected source {remoteEndPoint}; expected {_sourceFilter.Description}; dropped={_droppedFrames}");
        }
    }

    private static string PreviewHex(byte[] bytes)
    {
        const int maxPreviewBytes = 64;
        var previewLength = Math.Min(bytes.Length, maxPreviewBytes);
        var hex = Convert.ToHexString(bytes.AsSpan(0, previewLength));
        return bytes.Length > maxPreviewBytes ? $"{hex}..." : hex;
    }
}

public sealed class UdpKmSourceFilter
{
    public static UdpKmSourceFilter Disabled { get; } = new(null, Array.Empty<IPAddress>());

    private readonly HashSet<IPAddress> _allowedAddresses;

    private UdpKmSourceFilter(string? host, IReadOnlyCollection<IPAddress> allowedAddresses)
    {
        Host = host;
        _allowedAddresses = allowedAddresses.Select(Normalize).ToHashSet();
    }

    public string? Host { get; }
    public bool Enabled => _allowedAddresses.Count > 0;
    public string Description => Enabled
        ? $"{Host} ({string.Join(", ", _allowedAddresses.Select(address => address.ToString()).Order())})"
        : "any source";

    public static UdpKmSourceFilter FromHost(string? host)
    {
        var trimmed = host?.Trim();
        if (string.IsNullOrWhiteSpace(trimmed))
            return Disabled;

        var addresses = Dns.GetHostAddresses(trimmed);
        return addresses.Length == 0
            ? throw new SocketException((int)SocketError.HostNotFound)
            : new UdpKmSourceFilter(trimmed, addresses);
    }

    public static UdpKmSourceFilter FromAddresses(string host, params IPAddress[] addresses)
    {
        return addresses.Length == 0
            ? throw new ArgumentException("At least one address is required.", nameof(addresses))
            : new UdpKmSourceFilter(host, addresses);
    }

    public bool Allows(IPAddress remoteAddress)
    {
        return !Enabled || _allowedAddresses.Contains(Normalize(remoteAddress));
    }

    private static IPAddress Normalize(IPAddress address)
    {
        return address.IsIPv4MappedToIPv6 ? address.MapToIPv4() : address;
    }
}

File: mac/Core/Network/UdpKmStream.swift
import Foundation
import Network

/// UDP socket carrying encoded KmFrame events.
/// TODO: wrap frames in AEAD once TlsSession exposes exporter-derived key material.
public final class UdpKmStream {
    public enum StreamError: Error, Equatable {
        case invalidPort(UInt16)
    }

    public var onFrame: ((KmFrame) -> Void)?
    public var onFrameFrom: ((KmFrame, NWEndpoint) -> Void)?

    private let queue: DispatchQueue
    private var listener: NWListener?
    private var connections: [String: NWConnection] = [:]

    public init(queue: DispatchQueue = DispatchQueue(label: "com.barelyreal.udp-km")) {
        self.queue = queue
    }

    public func bind(localPort: UInt16) throws {
        guard let port = NWEndpoint.Port(rawValue: localPort) else {
            throw StreamError.invalidPort(localPort)
        }

        close()

        let listener = try NWListener(using: .udp, on: port)
        listener.newConnectionHandler = { [weak self] connection in
            self?.startInbound(connection)
        }
        listener.start(queue: queue)
        self.listener = listener
    }

    public func send(_ frame: KmFrame, to peer: NWEndpoint) {
        let payload = KmFrameCodec.encode(frame)
        queue.async { [weak self] in
            guard let self else { return }
            let connection = self.connection(to: peer)
            connection.send(content: payload, completion: .contentProcessed { _ in })
        }
    }

    public func close() {
        listener?.cancel()
        listener = nil

        for connection in connections.values {
            connection.cancel()
        }
        connections.removeAll()
    }

    private func startInbound(_ connection: NWConnection) {
        let key = String(describing: connection.endpoint)
        connections[key] = connection
        connection.start(queue: queue)
        receiveNext(on: connection)
    }

    private func receiveNext(on connection: NWConnection) {
        connection.receiveMessage { [weak self, weak connection] content, _, _, error in
            guard let self, let connection else { return }

            if let content {
                do {
                    let frame = try KmFrameCodec.decode(content)
                    self.onFrame?(frame)
                    self.onFrameFrom?(frame, connection.endpoint)
                } catch {
                    // Malformed UDP datagrams are dropped. Control channel will own reconnect policy.
                }
            }

            if error == nil {
                self.receiveNext(on: connection)
            }
        }
    }

    private func connection(to peer: NWEndpoint) -> NWConnection {
        let key = String(describing: peer)
        if let connection = connections[key] {
            return connection
        }

        let connection = NWConnection(to: peer, using: .udp)
        connections[key] = connection
        connection.start(queue: queue)
        return connection
    }
}
